### PR TITLE
[camera] Dispose resources correctly on setDescription

### PR DESCRIPTION
**Pull request by BradenBagby**
setDescription for CameraController was not correctly handling resources. It would recreate the _deviceOrientationSubscription each time it was called, which caused the subscription to not be disposed correctly. It also was not disposing of the previous device camera

*List which issues are fixed by this PR. You must list at least one issue.*

Issue seen here: https://github.com/flutter/flutter/issues/126823

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by jellynoone** [comment]
Hi, I was the one that reported the issue. Thank you for fixing on this!

Are you sure doing `await _initCalled` actually does anything, from the looks of the class, `_initCalled` might be marked as FutureOr<bool>? but its only ever `bool or null`.  _(I think this fact also puts into question the dispose method of this class which uses the same logic, the `_initCalled` is never really awaited for since its never a Future, so if the initialisation process is going on it will be null not Future<bool>)_

Additionally, what if a camera preview is being displayed and the camera is disposed before the new one is initialised? I don't know this is why I'm asking.

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (bparrishMines) [inline packages/camera/camera/lib/src/camera_controller.dart:301]: I don't like the idea that the `CameraController` calls `dispose` and is still remains usable. This doesn't seem like intended behavior. And now that I look at `setDescription`, is it possible to call `setDescription` instead of `initialize()` to initialize a camera? This also doesn't seem intended. 

I think it makes more sense to put this line in `setDescription` and not have this conditional.

**Comment by stuartmorgan** [comment]
@BradenBagby Are you still planning on updating this PR based on the feedback above?

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (BradenBagby) [reply packages/camera/camera/lib/src/camera_controller.dart:301]: Ill change and test it with that under setDescription. 

I agree its weird that you could call setDescription to completely initialize the controller. Since setDescription is really just handling what you used to have to do manually when changing cameras -> re-initialize a controller with a new description. We could throw calls to setDescription if initialize has never been called?

**Comment by BradenBagby** [comment]
Ive moved it to setDescription @bparrishMines . Let me know if you want setDescription to throw an error if its not already initialized instead of just initializing it for you. If thats the case, wouldn't need the dispose call within the if block, but we would probably still want to await _initCalled incase the user is spamming the switch camera button

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (bparrishMines) [inline packages/camera/camera/lib/src/camera_controller.dart:409]: I think this should be awaited. This looks like it could cause race conditions with `_initializeWithDescription` if unawaited.

File: packages/camera/camera/CHANGELOG.md
Comment (bparrishMines) [inline packages/camera/camera/CHANGELOG.md:3]: This could be more descriptive. I would use a sentence to explain the error that is being fixed. This PR also seems to fix 2 bugs, so I would include a sentence for each.

**Review by bparrishMines — COMMENTED** [review]
I added a couple more comments. 

As a side note, I didn't realize before that `setDescription` reinitializes the `CameraController`. I don't think the class was designed to be able to handle this, so we are starting to see bugs like this pop up. It may be worth evaluating whether the `setDescription` should only be used while recording for the short term. This doesn't block this PR from fixing these bugs though.

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (bparrishMines) [reply packages/camera/camera/lib/src/camera_controller.dart]: I agree. I think we should add the `_throwIfNotInitialized` method to the top of `setDescription`.

**Comment by BradenBagby** [comment]
Yes I also think it should just be setDescriptionWhileRecording. I should hopefully be able to get to fixing this pr in the next couple days

File: packages/camera/camera/CHANGELOG.md
Comment (bparrishMines) [inline packages/camera/camera/CHANGELOG.md:4]:
```suggestion
* Fixes bug where old camera resources were not disposed when switching between camera descriptions.
* Fixes bug where _deviceOrientationSubscription was recreated every time the camera description was changed.
```

**Review by bparrishMines — COMMENTED** [review]
This looks good to me except a missing test. Specifically, you could test that dispose is called in `setDescription`.

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (bparrishMines) [reply packages/camera/camera/lib/src/camera_controller.dart]: Are you going to add `_throwIfNotInitialized` to the beginning of the method?

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (bparrishMines) [reply packages/camera/camera/lib/src/camera_controller.dart]: I'm not sure of the implications of this either.  I need to manually test it out to see.  We may have made a mistake with adding this feature with the current version of the plugin. Since we have to call dispose on the camera anyway now, there is no benefit in using this method over creating a new `CameraController` besides adding more complexity to the plugin  (except when recording). Calling `setDescription` might also just require a following `setState` call so that `buildPreview` can be called with the new `cameraId`.

**Comment by stuartmorgan** [comment]
Is this ready for a second review? (It looks like it's failing analysis, but it's not clear to me from discussion above if it's ready for a second look more generally.)

**Comment by stuartmorgan** [comment]
Update from triage: @bparrishMines will add tests to this when time allows.

**Comment by bparrishMines** [comment]
@stuartmorgan I added the tests and also updated `_initCalled` with a field that actually waits for initialization.

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Review by bparrishMines — APPROVED** [review]
LGTM

**Pull request merged**
